File: src/modules/Workspaces/WorkspacesCsharpLibrary/SettingsService/PTSettingsClient.cs
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using System.IO;
using System.IO.Pipes;
using System.Security.Principal;

namespace WorkspacesCsharpLibrary.SettingsService;

/// <summary>
/// Managed client for PTSettingsSvc.  Mirrors the native client and the
/// service's wire protocol (see Protocol.h) so the Editor, runner and unit
/// tests can talk to the service without P/Invoke.  The service treats the
/// payload as opaque bytes; all JSON / schema concerns live in the caller.
/// </summary>
public static class PTSettingsClient
{
    /// <summary>Coarse result surfaced to callers (mirrors the service status bands).</summary>
    public enum Result : byte
    {
        /// <summary>Request succeeded.</summary>
        Ok = 0,

        /// <summary>GetBlob: the blob does not exist yet (service is up).</summary>
        NotFound,

        /// <summary>Caller authentication / namespace check failed.</summary>
        AuthRejected,

        /// <summary>No service to talk to (not installed / not running).</summary>
        Unavailable,

        /// <summary>Framing / unexpected protocol error.</summary>
        Protocol,

        /// <summary>Underlying file IO failed in the service.</summary>
        IoError,
    }

    // Mirror of PTSettingsSvc pipe naming (Approach 4 / §12.8): each user has
    // their own service instance PTSettingsSvc_<SID>, reachable at
    // \\.\pipe\PTSettingsSvc_<SID>.  We derive <SID> from OUR OWN token so we
    // always reach our own user's instance.  NamedPipeClientStream takes the
    // name without the \\.\pipe\ prefix.  Computed lazily and cached.
    private const string PipeNamePrefix = "PTSettingsSvc_";

    private static readonly Lazy<string> _pipeName = new(() =>
        PipeNamePrefix + (WindowsIdentity.GetCurrent().User?.Value ?? string.Empty));

    public static string PipeName => _pipeName.Value;

    // Mirror of PTSettingsSvc::kMaxPayloadBytes (1 MiB).
    private const int MaxPayloadBytes = 1 * 1024 * 1024;

    private const int ConnectTimeoutMs = 3000;

    // Opcodes (mirror of PTSettingsSvc::Opcode).
    private const byte OpPing = 0x00;
    private const byte OpGetBlob = 0x01;
    private const byte OpPutBlob = 0x02;

    /// <summary>Liveness probe.  Authentication still runs server-side.</summary>
    public static Result Ping()
    {
        return RoundTrip(OpPing, ReadOnlySpan<byte>.Empty, out _);
    }

    /// <summary>Reads this caller's namespace blob.  Returns NotFound if none exists yet.</summary>
    public static Result GetBlob(out byte[] blob)
    {
        var rc = RoundTrip(OpGetBlob, ReadOnlySpan<byte>.Empty, out var resp);
        blob = rc == Result.Ok ? resp : Array.Empty<byte>();
        return rc;
    }

    /// <summary>Atomically replaces this caller's namespace blob with the given bytes.</summary>
    public static Result PutBlob(ReadOnlySpan<byte> blob)
    {
        return RoundTrip(OpPutBlob, blob, out _);
    }

    private static Result RoundTrip(byte opcode, ReadOnlySpan<byte> payload, out byte[] response)
    {
        response = Array.Empty<byte>();
        if (payload.Length > MaxPayloadBytes)
        {
            return Result.Protocol;
        }

        NamedPipeClientStream pipe;
        try
        {
            // TokenImpersonation lets the service impersonate us to read our
            // SID (per-user data partitioning) and open our process for the
            // image-path / signature checks.
            pipe = new NamedPipeClientStream(
                ".",
                PipeName,
                PipeDirection.InOut,
                PipeOptions.None,
                TokenImpersonationLevel.Impersonation);
            pipe.Connect(ConnectTimeoutMs);
        }
        catch (TimeoutException)
        {
            return Result.Unavailable;
        }
        catch (IOException)
        {
            return Result.Unavailable;
        }
        catch (UnauthorizedAccessException)
        {
            return Result.Unavailable;
        }

        using (pipe)
        {
            try
            {
                Span<byte> header = stackalloc byte[5];
                header[0] = opcode;
                BitConverter.TryWriteBytes(header[1..], (uint)payload.Length);
                pipe.Write(header);
                if (payload.Length > 0)
                {
                    pipe.Write(payload);
                }

                pipe.Flush();

                Span<byte> respHeader = stackalloc byte[5];
                if (!ReadExact(pipe, respHeader))
                {
                    return Result.Protocol;
                }

                byte status = respHeader[0];
                uint respLen = BitConverter.ToUInt32(respHeader[1..]);
                if (respLen > MaxPayloadBytes)
                {
                    return Result.Protocol;
                }

                if (respLen > 0)
                {
                    response = new byte[respLen];
                    if (!ReadExact(pipe, response))
                    {
                        response = Array.Empty<byte>();
                        return Result.Protocol;
                    }
                }

                return MapStatus(status);
            }
            catch (IOException)
            {
                return Result.Protocol;
            }
        }
    }

    private static bool ReadExact(Stream stream, Span<byte> dest)
    {
        int offset = 0;
        while (offset < dest.Length)
        {
            int got = stream.Read(dest[offset..]);
            if (got <= 0)
            {
                return false;
            }

            offset += got;
        }

        return true;
    }

    private static Result MapStatus(byte status)
    {
        // Mirror of PTSettingsSvc::Status, collapsed to the coarse Result.
        return status switch
        {
            0x00 => Result.Ok,
            0x20 => Result.NotFound,
            0x10 or 0x11 or 0x12 => Result.AuthRejected,
            0x21 => Result.IoError,
            _ => Result.Protocol,
        };
    }
}

File: src/modules/Workspaces/WorkspacesSettingsClient/PTSettingsClient.cpp
// Copyright (c) Microsoft Corporation
// Licensed under the MIT license.

#include "PTSettingsClient.h"
#include "../WorkspacesSettingsService/protocol/Protocol.h"
#include "../WorkspacesSettingsService/protocol/PipeName.h"

#include <windows.h>
#include <vector>
#include <cstring>
#include <string>

namespace PTSettingsClient
{
    namespace
    {
        using PTSettingsSvc::kMaxPayloadBytes;
        using PTSettingsSvc::Opcode;
        using PTSettingsSvc::Status;

        // This client reaches ITS OWN user's service instance, whose pipe is
        // \\.\pipe\PTSettingsSvc_<SID> where <SID> is our own token SID
        //.  Computed once per process.
        const std::wstring& OwnPipeName()
        {
            static const std::wstring name =
                PTSettingsSvc::BuildPipeName(PTSettingsSvc::CurrentProcessUserSidString());
            return name;
        }

        struct PipeHandle
        {
            HANDLE h = INVALID_HANDLE_VALUE;
            ~PipeHandle()
            {
                if (h != INVALID_HANDLE_VALUE) CloseHandle(h);
            }
        };

        bool Connect(PipeHandle& out)
        {
            const std::wstring& pipeName = OwnPipeName();
            for (int attempt = 0; attempt < 3; ++attempt)
            {
                HANDLE h = CreateFileW(pipeName.c_str(),
                                       GENERIC_READ | GENERIC_WRITE,
                                       0,
                                       nullptr,
                                       OPEN_EXISTING,
                                       // Allow the server to impersonate us
                                       // so it can read our SID; anything
                                       // weaker yields an Anonymous token
                                       // and the server's auth check fails.
                                       SECURITY_SQOS_PRESENT | SECURITY_IMPERSONATION,
                                       nullptr);
                if (h != INVALID_HANDLE_VALUE)
                {
                    out.h = h;
                    return true;
                }
                DWORD err = GetLastError();
                if (err != ERROR_PIPE_BUSY && err != ERROR_FILE_NOT_FOUND)
                {
                    return false;
                }
                WaitNamedPipeW(pipeName.c_str(), 2000);
            }
            return false;
        }

        bool WriteAll(HANDLE h, const void* buf, DWORD len)
        {
            const BYTE* p = static_cast<const BYTE*>(buf);
            while (len > 0)
            {
                DWORD wrote = 0;
                if (!WriteFile(h, p, len, &wrote, nullptr) || wrote == 0) return false;
                p += wrote;
                len -= wrote;
            }
            return true;
        }

        bool ReadAll(HANDLE h, void* buf, DWORD len)
        {
            BYTE* p = static_cast<BYTE*>(buf);
            while (len > 0)
            {
                DWORD got = 0;
                if (!ReadFile(h, p, len, &got, nullptr) || got == 0) return false;
                p += got;
                len -= got;
            }
            return true;
        }

        Result MapStatus(Status s)
        {
            switch (s)
            {
            case Status::Ok:               return Result::Ok;
            case Status::AuthFailToken:
            case Status::AuthFailCaller:   return Result::AuthRejected;
            case Status::NamespaceUnknown: return Result::NamespaceUnknown;
            case Status::BadRequest:
            case Status::UnknownOpcode:    return Result::ProtocolError;
            case Status::PayloadTooLarge:  return Result::PayloadTooLarge;
            case Status::NotFound:         return Result::NotFound;
            case Status::IoError:          return Result::IoError;
            }
            return Result::UnknownStatus;
        }

        Result RoundTrip(Opcode op, const void* payload, uint32_t payloadLen,
                         std::vector<uint8_t>& outResp)
        {
            outResp.clear();
            if (payloadLen > kMaxPayloadBytes)
            {
                return Result::PayloadTooLarge;
            }

            PipeHandle pipe;
            if (!Connect(pipe))
            {
                return Result::ServiceUnavailable;
            }

            uint8_t opByte = static_cast<uint8_t>(op);
            if (!WriteAll(pipe.h, &opByte, sizeof(opByte)) ||
                !WriteAll(pipe.h, &payloadLen, sizeof(payloadLen)) ||
                (payloadLen > 0 && !WriteAll(pipe.h, payload, payloadLen)))
            {
                return Result::ProtocolError;
            }

            uint8_t statusByte = 0;
            uint32_t respLen = 0;
            if (!ReadAll(pipe.h, &statusByte, sizeof(statusByte)) ||
                !ReadAll(pipe.h, &respLen, sizeof(respLen)))
            {
                return Result::ProtocolError;
            }
            if (respLen > kMaxPayloadBytes)
            {
                return Result::ProtocolError;
            }
            if (respLen > 0)
            {
                outResp.resize(respLen);
                if (!ReadAll(pipe.h, outResp.data(), respLen))
                {
                    outResp.clear();
                    return Result::ProtocolError;
                }
            }
            return MapStatus(static_cast<Status>(statusByte));
        }
    }

    Result Ping()
    {
        std::vector<uint8_t> resp;
        return RoundTrip(Opcode::Ping, nullptr, 0, resp);
    }

    Result GetBlob(std::vector<uint8_t>& outBytes)
    {
        return RoundTrip(Opcode::GetBlob, nullptr, 0, outBytes);
    }

    Result PutBlob(const std::vector<uint8_t>& bytes)
    {
        std::vector<uint8_t> resp;
        return RoundTrip(Opcode::PutBlob,
                         bytes.data(),
                         static_cast<uint32_t>(bytes.size()),
                         resp);
    }
}
